rpc/resource/example_test.go

```diff
@@ -0,0 +1,136 @@
+package resource_test
+
+import (
+	"fmt"
+
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/resource"
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/resource/external"
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/resource/resourcepb"
+	"github.com/infobloxopen/atlas-app-toolkit/rpc/resource/uuid"
+)
+
+type ExampleGoType struct {
+	ID         resource.Identifier
+	ExternalID resource.Identifier
+	VarName    string
+}
+
+type ExampleProtoMessage struct {
+	Id         *resourcepb.Identifier
+	ExternalId *resourcepb.Identifier
+	VarName    string
+}
+
+func (ExampleProtoMessage) XXX_MessageName() string { return "ExampleProtoMessage" }
+func (ExampleProtoMessage) Reset()                  {}
+func (ExampleProtoMessage) String() string          { return "ExampleProtoMessage" }
+func (ExampleProtoMessage) ProtoMessage()           {}
+
+func Example() {
+	// register external codec for default resources
+	resource.RegisterCodec(external.NewCodec(), nil)
+	// register uuid codec for TestProtoMessage resources
+	resource.RegisterCodec(uuid.NewCodec("simpleapp", "examples"), &ExampleProtoMessage{})
+
+	// and you want to convert PB type to your application type
+	toGoTypeFunc := func(msg *ExampleProtoMessage) (*ExampleGoType, error) {
+		var v ExampleGoType
+
+		// arbitrary variables
+		v.VarName = msg.VarName
+
+		// convert RPC identifier using UUID Codec for ExampleProtoMessage resource type
+		if id, err := resource.Decode(msg, msg.Id); err != nil {
+			return nil, err
+		} else {
+			v.ID = id
+		}
+		// convert RPC identifier using External Codec for default resource type
+		if id, err := resource.Decode(nil, msg.ExternalId); err != nil {
+			return nil, err
+		} else {
+			v.ExternalID = id
+		}
+		return &v, nil
+	}
+
+	// let's create PB message
+	pb := &ExampleProtoMessage{
+		Id: &resourcepb.Identifier{
+			// application name is empty because Id field stores internal identifier
+			ApplicationName: "",
+			// resource type is empty as well
+			ResourceType: "",
+			// resource id is empty because we want to create new TestProtoMessage
+			ResourceId: "",
+		},
+		// ExternalId stores data about "external_resource" that belongs to
+		// "externalapp" and has id "1"
+		ExternalId: &resourcepb.Identifier{
+			ApplicationName: "externalapp",
```

Comment:
so is the idea that these reference representations should only be used when you're storing any kind of object from any king of application/resource? 

If this example app has a reference to a field that is for a specific application, then the `ApplicationName` shouldn't be allowed to be entered by users. In that case, they can just use a `string` or `int` field to represent it. Correct?